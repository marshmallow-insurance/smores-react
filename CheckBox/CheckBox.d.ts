import React, { ReactNode } from 'react';
export type CheckBoxProps = {
    id?: string;
    checked: boolean;
    children: ReactNode;
    toggle: () => void;
    error?: boolean;
    errorMsg?: string;
};
export declare const CheckBox: React.ForwardRefExoticComponent<CheckBoxProps & React.RefAttributes<HTMLInputElement>>;
