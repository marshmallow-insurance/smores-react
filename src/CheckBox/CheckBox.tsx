import React, { forwardRef, ReactNode } from 'react'
import styled from 'styled-components'
import { focusOutline } from '../utils/focusOutline'

import { Text } from '../Text'
import { theme } from '../theme'
import { useUniqueId } from '../utils/id'

export type CheckBoxProps = {
  id?: string
  checked: boolean
  children: ReactNode
  toggle: () => void
  error?: boolean
  errorMsg?: string
}

export const CheckBox = forwardRef<HTMLInputElement, CheckBoxProps>(
  function CheckBox(
    { id: idProp, checked, children, toggle, error, errorMsg },
    ref,
  ) {
    const id = useUniqueId(idProp)
    return (
      <>
        <BoxContainer id={id}>
          <Text
            tag="span"
            typo="base"
            color={error ? 'strawberry' : 'liquorice'}
          >
            {children}
          </Text>

          <input
            ref={ref}
            type="checkbox"
            checked={checked}
            onChange={toggle}
          />
          <Checkmark $error={error} />
        </BoxContainer>
        {error && errorMsg && <ErrorBox>{errorMsg}</ErrorBox>}
      </>
    )
  },
)

const Checkmark = styled.span<{ $error?: boolean }>`
  position: absolute;
  left: 0;
  width: 24px;
  height: 24px;
  border: ${({ $error }) =>
    $error
      ? `solid 2px ${theme.colors.strawberry}`
      : `solid 2px ${theme.colors.liquorice}`};
  box-sizing: border-box;
  border-radius: 1px;

  &:before {
    content: '';
    position: absolute;
    display: none;
    top: 9px;
    left: 5px;
    width: 3px;
    height: 8px;
    border-radius: 4px;
    background-color: ${theme.colors.cream};
    -webkit-transform: rotate(316deg);
    -ms-transform: rotate(316deg);
    transform: rotate(316deg);
  }

  &:after {
    content: '';
    position: absolute;
    display: none;
    top: 3px;
    left: 11px;
    width: 3px;
    height: 15px;
    border-radius: 4px;
    background-color: ${theme.colors.cream};
    -webkit-transform: rotate(43deg);
    -ms-transform: rotate(43deg);
    transform: rotate(43deg);
  }
`

const BoxContainer = styled.label`
  position: relative;
  display: flex;
  align-items: flex-start;
  padding-left: 32px;
  user-select: none;
  cursor: pointer;

  input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    background-color: ${theme.colors.cream};

    &:checked ~ ${Checkmark} {
      background-color: ${theme.colors.liquorice};
      border: solid 2px ${theme.colors.liquorice};
    }

    &:checked ~ ${Checkmark}:before {
      display: block;
    }

    &:checked ~ ${Checkmark}:after {
      display: block;
    }

    ${focusOutline({ selector: `&:focus-visible ~ ${Checkmark}` })}
  }

  &:hover {
    ${Checkmark} {
      background-color: ${theme.colors.coconut};
      border: solid 2px ${theme.colors.liquorice};
    }
  }

  @media (min-width: 768px) {
    padding-left: 32px;
  }

  span {
    border-radius: 6px;
  }
`

const ErrorBox = styled.span`
  padding-top: 8px;
  font-size: 12px;
  color: ${theme.colors.strawberry};
`
