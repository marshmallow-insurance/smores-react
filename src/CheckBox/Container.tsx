import React, { useState } from 'react'
import { CheckBox } from './CheckBox'

export const Container = () => {
  const [checked, setChecked] = useState(false)

  return (
    <CheckBox
      id="agreement"
      checked={checked}
      toggle={() => setChecked(!checked)}
    >
      I agree
    </CheckBox>
  )
}
