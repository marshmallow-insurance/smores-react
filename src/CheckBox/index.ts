export { CheckBox, CheckBoxProps } from './CheckBox'
