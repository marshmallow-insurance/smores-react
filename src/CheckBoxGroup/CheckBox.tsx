import React, { FC, ReactNode } from 'react'
import styled from 'styled-components'

import { Text } from '../Text'
import { theme } from '../theme'

type Props = {
  id: string
  checked: boolean
  children: ReactNode
  toggle: () => void
}

export const CheckBox: FC<Props> = ({ id, checked, children, toggle }) => (
  <BoxContainer id={id}>
    <Text tag="span" typo="base">
      {children}
    </Text>

    <input type="checkbox" checked={checked} onChange={toggle} />
    <Checkmark />
  </BoxContainer>
)

const Checkmark = styled.span`
  position: absolute;
  left: 0;
  width: 24px;
  height: 24px;
  border: solid 1px ${theme.colors.secondary};
  box-sizing: border-box;
  border-radius: 1px;

  &:after {
    content: '';
    position: absolute;
    display: none;
    top: 3px;
    left: 7px;
    width: 5px;
    height: 12px;
    border: solid white;
    border-width: 0 2px 2px 0;
    -webkit-transform: rotate(45deg);
    -ms-transform: rotate(45deg);
    transform: rotate(45deg);
  }
`

const BoxContainer = styled.label`
  position: relative;
  display: flex;
  align-items: flex-start;
  padding-left: 32px;
  user-select: none;
  cursor: pointer;

  input {
    position: absolute;
    opacity: 0;
    cursor: pointer;

    &:checked ~ ${Checkmark} {
      background-color: ${theme.colors.secondary};
      border: solid 1px ${theme.colors.secondary};
    }

    &:checked ~ ${Checkmark}:after {
      display: block;
    }
  }

  &:hover {
    ${Checkmark} {
      background-color: ${theme.colors.background};
      border: solid 1px ${theme.colors.secondary};
    }
  }

  @media (min-width: 768px) {
    padding-left: 32px;
  }
`
