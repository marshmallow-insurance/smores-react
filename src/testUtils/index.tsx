import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { SmoresThemeProvider } from '../ThemeProvider/ThemeProvider'

const Providers = ({ children }: { children: React.ReactNode }) => {
  return <SmoresThemeProvider>{children}</SmoresThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'
export * from '@testing-library/jest-dom/vitest'
export * from 'jest-styled-components'

export { customRender as render }
