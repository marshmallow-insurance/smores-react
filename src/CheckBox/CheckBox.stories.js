import React from 'react'
import { CheckBox } from './CheckBox'
import { Container as CheckBoxContainer } from './Container'

export default {
  title: 'Checkbox',
  component: CheckBox,
}

const ContainerTemplate = (args) => <CheckBoxContainer {...args} />

export const Default = ContainerTemplate.bind({})

const Template = (args) => <CheckBox {...args} />

export const WithError = Template.bind({})

WithError.args = {
  children: 'I disagree',
  error: true,
}

export const WithErrorLabel = Template.bind({})

WithErrorLabel.args = {
  children: 'I disagree',
  error: true,
  errorMsg: 'Something has gone wrong',
}
