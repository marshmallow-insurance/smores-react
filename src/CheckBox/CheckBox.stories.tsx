import React from 'react'
import { CheckBox, CheckBoxProps } from './CheckBox'
import { Container as CheckBoxContainer } from './Container'

export default {
  title: 'Checkbox',
  component: CheckBox,
}

const ContainerTemplate = () => <CheckBoxContainer />

export const Default = ContainerTemplate.bind({})

const Template = (props: CheckBoxProps) => <CheckBox {...props} />

export const WithError = Template.bind({})

WithError.args = {
  children: 'I disagree',
  error: true,
}

export const WithErrorLabel = Template.bind({})

WithErrorLabel.args = {
  children: 'I disagree',
  error: true,
  errorMsg: 'Something has gone wrong',
}
