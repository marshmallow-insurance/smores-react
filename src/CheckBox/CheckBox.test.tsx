import React from 'react'
import { expect, it } from 'vitest'
import { render } from '../testUtils'
import { CheckBox } from './CheckBox'

describe('CheckBox', () => {
  it('renders correctly when unchecked', () => {
    const { container } = render(
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      <CheckBox checked={false} toggle={() => {}}>
        Accept Terms
      </CheckBox>,
    )
    expect(container).toMatchSnapshot()
  })
  it('renders correctly when checked', () => {
    const { container } = render(
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      <CheckBox checked={true} toggle={() => {}}>
        Accept Terms
      </CheckBox>,
    )
    expect(container).toMatchSnapshot()
  })
  it('renders correctly with an error state', () => {
    const { container } = render(
      <CheckBox
        checked={false}
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        toggle={() => {}}
        error={true}
        errorMsg="You must accept the terms"
      >
        Accept Terms
      </CheckBox>,
    )
    expect(container).toMatchSnapshot()
  })
})
