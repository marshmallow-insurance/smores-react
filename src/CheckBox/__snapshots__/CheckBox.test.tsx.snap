// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CheckBox > renders correctly when checked 1`] = `
<div>
  <label
    class="sc-hjsuWn iyKZva"
    id="MM_SMORES_1"
  >
    <span
      class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
      cursor="inherit"
      title=""
    >
      Accept Terms
    </span>
    <input
      checked=""
      type="checkbox"
    />
    <span
      class="sc-jwTyAe kYeyOg"
    />
  </label>
</div>
`;

exports[`CheckBox > renders correctly when unchecked 1`] = `
<div>
  <label
    class="sc-hjsuWn iyKZva"
    id="MM_SMORES_0"
  >
    <span
      class="sc-bRKDuR bHjpxP sc-dTvVRJ JyLYT"
      cursor="inherit"
      title=""
    >
      Accept Terms
    </span>
    <input
      type="checkbox"
    />
    <span
      class="sc-jwTyAe kYeyOg"
    />
  </label>
</div>
`;

exports[`CheckBox > renders correctly with an error state 1`] = `
<div>
  <label
    class="sc-hjsuWn iyKZva"
    id="MM_SMORES_2"
  >
    <span
      class="sc-bRKDuR bHjpxP sc-dTvVRJ gXSYMm"
      cursor="inherit"
      title=""
    >
      Accept Terms
    </span>
    <input
      type="checkbox"
    />
    <span
      class="sc-jwTyAe iuKYMp"
    />
  </label>
  <div
    class="sc-jJLAfE llFNkS"
  >
    You must accept the terms
  </div>
</div>
`;
